### PR TITLE
MAINT: trapz works on non-monotonic x

### PR DESCRIPTION
**Pull request by andyfaff**
This PR allows `trapz` to take non monotonic `x` arrays. It achieves this by a sort along the required axis.
1 and 3-D tests are added.

### CODE REVIEW

**Comment by andyfaff** [comment]
Think the error comes because the x arr needs to be broadcast against y.

**Comment by andyfaff** [comment]
Only one test failing now, but I'm not sure how to use `broadcast_to` with masked arrays, unless I do some object testing.

File: numpy/lib/function_base.py [hunk 1]
```diff
@@ -3712,19 +3712,34 @@ def trapz(y, x=None, dx=1.0, axis=-1):
...
             shape = [1]*y.ndim
-            shape[axis] = d.shape[0]
-            d = d.reshape(shape)
-        else:
-            d = diff(x, axis=axis)
+            shape[axis] = x.shape[0]
+            x = np.copy(x).reshape(shape)
+
+        x, y = np.broadcast_arrays(x, y)
+        idxs = _argsort_indices(x, axis)
```
[the remark this answers is not in the document]
Comment (eric-wieser) [reply numpy/lib/function_base.py]: Apologies, I muddled this with the permutation PR (#9880), which at one point used a similar name

**Comment by andyfaff** [comment]
The rationale for `trapz` is that it's for sampled x/y points. Obviously the order matters.

```
>>> np.random.seed(2)
>>> x = np.arange(6)
>>> np.random.shuffle(x)
>>> print(np.trapz(x,x))
-8.0
```

There will be a non-negligible population who think the answer should be 0.5 * 5 * 5 = 12.5.

I don't mind if the PR doesn't progress further, but then the documentation should probably be changed to say that the order of x matters.

**Comment by eric-wieser** [comment]
So to elaborate, the current behavior allows integration of parametric functions that are not monotonic in `x` and/or `y`, in a mathematically meaningful way.

If your data is un-ordered, then I think it's up to you as the user to put it in the order that makes sense for your function.

Another example of why the existing implementation is useful:
```
>>> theta = np.linspace(0, 2*np.pi, 1000)
>>> x = np.sin(theta)
>>> y = np.cos(theta)
>>> np.trapz(y, x)
3.1415719413758412   # area of the approximate circle = approximation to pi!
```

It does sound like the documentation could be improved

**Comment by eric-wieser** [comment]
> There will be a non-negligible population who think the answer should be 0.5 * 5 * 5 = 12.5.

These people will probably realize that they need to order their data when they try to do a line plot in matplotlib, which also cares about the order of their data.

More worryingly, if you organically have unordered data (random measurements from a process), sorting by one coordinate is a very bad way to produce a fit, and integration under that fit doesn't sound like a meaningful thing to do to me.

**Pull request closed without merging**